### PR TITLE
WIP: Remix Example

### PR DESCRIPTION
**Pull request by ScriptedAlchemy**
Work in progress, remix federation. 

Plugin setup needs to be abstracted away as this is lower level than end user should experience 
TODO: 
- [x] Move plugin stuff to module-federation/universe(repo) and module-federation/enhanced(package)
- - [ ] clean up this pr
- [ ] Create module-federation/remix if all aspects of federation cannot be internally integrated in a universal manner
- [ ] Support ESM outputs in module-federation/node (package) - CommonJS works fine for now, via "async-node" target.
- [ ] Move Dependency Hoisting Plugin into Module Federation Plugin to support single runtime chunk universally 
- [ ] Pull more complex example off the shelf and test all the parts actually work as it appears. 
- [x] Make Rsbuild plugin for Rspack: https://github.com/ScriptedAlchemy/remix

### CODE REVIEW

File: remix/app1/webpack.browser.js [hunk 1]
```diff
@@ -0,0 +1,207 @@
+import * as path from "node:path";
+
+import {readConfig} from "@remix-run/dev/dist/config.js";
+import {EsbuildPlugin} from "esbuild-loader";
+import {toManifest, writeManifest} from "./utils/manifest.js";
+import {default as Enhanced} from '@module-federation/enhanced'
+import DelegatesModulePlugin from '@module-federation/utilities/src/plugins/DelegateModulesPlugin.js'
+const {ModuleFederationPlugin, AsyncBoundaryPlugin} = Enhanced
+class HoistContainerReferences {
```
Comment (ScriptedAlchemy) [inline remix/app1/webpack.browser.js]: -Move this back into module-federation/enhanced
-Deprecate whatever im doing in Next.js to accomplish this. 
-Rewrite it again

File: remix/app1/webpack.browser.js [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        },
+        "@remix-run/react/": {
+          singleton: true
+        },
+        "@remix-run/": {
+          singleton: true
+        }
+      }
+    }),
+    {
```
Comment (ScriptedAlchemy) [inline remix/app1/webpack.browser.js]: Make rsbuild plugin

File: remix/app1/webpack.server.js [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+const mode =
+  process.env.NODE_ENV === "production" ? "production" : "development";
+const remixConfig = await readConfig();
+const isModule = remixConfig.serverModuleFormat === "esm";
+
+console.log({isModule});
+
+if (!isModule) {
+  fs.writeFileSync('./build/package.json', JSON.stringify({type: "commonjs"}))
```
Comment (ScriptedAlchemy) [inline remix/app1/webpack.server.js]: Hack to make common.js output work correctly in a esm input. 
Move to build pluign

File: remix/app1/webpack.server.js [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+  devtool: mode === "development" ? false : undefined,
+  target: 'async-node',
+  entry: remixConfig.serverEntryPoint
+    ? path.resolve(remixConfig.rootDirectory, remixConfig.serverEntryPoint)
+    : serverBuildModule,
+  experiments: isModule ? {outputModule: true} : undefined,
+  externalsType: isModule ? "module" : undefined,
+  externalsPresets: {node: true},
+  externals: [
+    nodeExternals({
```
Comment (ScriptedAlchemy) [inline remix/app1/webpack.server.js:44]: disable externals or opt shared modules and their dependencies out of externalization

File: remix/app1/webpack.server.js [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    filename: path.basename(remixConfig.serverBuildPath),
+    library: {type: isModule ? "module" : "commonjs"},
+    chunkFormat: isModule ? "module" : "commonjs",
+    chunkLoading: isModule ? "import" : undefined,
+    module: isModule,
+    path: path.dirname(remixConfig.serverBuildPath),
+    publicPath: remixConfig.publicPath,
+    assetModuleFilename: "_assets/[name]-[contenthash][ext]",
+    cssChunkFilename: "_assets/[name]-[contenthash][ext]",
+    chunkFilename: "[name]-[chunkhash].js",
```
Comment (ScriptedAlchemy) [inline remix/app1/webpack.server.js]: Fix in node federation. Chunks must always hash.
Long term, node federation chunk load logic should be able to tell if federated import and not perform fs lookup. 
Dev output causes fs lookup to be the same and remote loads same file name from host filesystem.

File: remix/app1/webpack.server.js [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+
+export default config;
+
+/**
+ *
+ * @param {import("@remix-run/dev").ResolvedRemixConfig} config
+ * @param {import("@remix-run/dev").AssetsManifest} manifest
+ * @returns
+ */
+function createServerBuildEntry(config, manifest) {
```
Comment (ScriptedAlchemy) [inline remix/app1/webpack.server.js]: Attach Module Federation v1.5 hooks here, perform distributed route discovery across remotes automatically. 
Easy to do, next.js made it impossible

File: remix/app2/utils/manifest.js [hunk 7]
```diff
@@ -0,0 +1,142 @@
+import * as fs from "node:fs";
```
Comment (ScriptedAlchemy) [inline remix/app2/utils/manifest.js:1]: make plugin

File: remix/app1/app/routes/_index.tsx [hunk 8]
```diff
@@ -0,0 +1,20 @@
+import { useLoaderData } from "@remix-run/react";
+import {lazy,Suspense}  from "react";
+import Button from 'app2/button'
+export function loader() {
+  return { message: "Hello, World!" };
+}
+// const Button = lazy(() => import('app2/button'));
+export default function Home() {
+    console.log('test');
```
Comment (fyodorovandrei) [inline remix/app1/app/routes/_index.tsx]:
```suggestion
```
😆

**Pull request merged**